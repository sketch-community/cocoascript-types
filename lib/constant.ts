export const NAMESPACE = 'cocoascript';

/**
 * 来自 Apple 文档的前缀
 */
export const APPLE_IDENTIFIER_PREFIX = 'doc://com.apple.documentation';

/**
 * 生成类型定义文件的前缀
 */
export const TS_FILE_IDENTIFIER_PREFIX = 'https://developer.apple.com';
