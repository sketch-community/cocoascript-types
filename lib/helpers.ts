import { NAMESPACE } from './constant';
import { shim } from './shim';
import { extend } from './extend';

export function capitalize(str: string) {
  return str[0].toUpperCase() + str.slice(1, str.length);
}

export function normalizeType(
  type: string,
  options: {
    extend?: boolean;
    withNamespace?: boolean;
  } = {
    extend: true,
    withNamespace: false,
  }
) {
  let finalType = type;
  if (options.extend && extend[finalType]) {
    return extend[finalType];
  }
  if (shim[finalType]) {
    return shim[finalType];
  }
  if (options.withNamespace) {
    return `${NAMESPACE}.${finalType}`;
  }
  return finalType;
}

export function getTokens(doc: any) {
  return doc.primaryContentSections?.[0]?.declarations?.[0]?.tokens || [];
}
