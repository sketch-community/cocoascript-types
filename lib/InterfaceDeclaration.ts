import * as fs from 'fs';
import * as path from 'path';
import { CodeGenerator } from './CodeGenerator';
import { NAMESPACE } from './constant';
import { normalizeType } from './helpers';
import { MethodDeclaration } from './MethodDeclaration';
import { PropertyDeclaration } from './PropertyDeclaration';
import { Token } from './types';

type InterfaceType = 'interface' | 'protocol';

export class InterfaceDeclaration {
  id: string;
  type: 'interface' | 'protocol';
  identifier: string;
  description: string;
  inheritsFrom = null;
  properties: Array<MethodDeclaration | PropertyDeclaration> = [];

  constructor(id: string, type: InterfaceType, identifier: string, description: string) {
    this.id = id;
    this.type = type;
    this.identifier = identifier;
    this.description = description;
  }

  static initFromTokens(id: string, doc: any, tokens: Token[]) {
    const idToken = tokens.find(t => t.kind === 'identifier')!;
    const decl = new InterfaceDeclaration(
      id,
      tokens[0].text.replace('@', '') as any,
      idToken.text,
      doc.abstract?.[0]?.text
    );
    if (doc.relationshipsSections) {
      const inheritsSection = doc.relationshipsSections.find((s: any) => s.type === 'inheritsFrom');
      if (inheritsSection) {
        const inheritsRef = doc.references[inheritsSection.identifiers[0]];
        decl.inheritsFrom = inheritsRef.title;
      }
    }
    if (doc.topicSections) {
      doc.topicSections.forEach((topic: any) => {
        topic.identifiers.forEach((id: string) => {
          const property = this.createIdentifier(decl, id);
          if (property) {
            decl.properties.push(property);
          }
        });
      });
    }
    return decl;
  }

  static createIdentifier(clazz: InterfaceDeclaration, id: string) {
    const docFile = path.resolve(id.replace('doc://com.apple.documentation/', '') + '.json');
    if (!fs.existsSync(docFile)) {
      return;
    }
    const doc = require(docFile);
    const firstSection = doc.primaryContentSections && doc.primaryContentSections[0];
    if (!firstSection) {
      return;
    }
    if (firstSection.kind !== 'declarations') {
      return;
    }
    const { tokens } = firstSection.declarations[0];
    let index = 0;
    if (tokens[index].text === '- (') {
      return MethodDeclaration.initFromTokens(doc.identifier.url, clazz, tokens);
    }
    if (tokens[index].text === '@property' || tokens[index].kind === 'typeIdentifier') {
      return PropertyDeclaration.initFromTokens(doc.identifier.url, tokens);
    }
  }

  generate() {
    const code = new CodeGenerator();
    code.appendLine(`declare namespace ${NAMESPACE} {`).indent();
    code.appendLine('/**');
    if (this.description) {
      code.appendLine(` * ${this.description}`);
    }
    code.appendLine(` * ${this.id}`);
    code.appendLine(' */');

    code.appendLine(`interface ${this.identifier}${this.inheritsFrom ? ` extends ${this.inheritsFrom} ` : ' '}{`).indent();
    if (this.type === 'interface') {
      const alloc = new MethodDeclaration('', this);
      alloc.identifiers.push({
        id: 'alloc',
      });
      alloc.returnType = 'instancetype';
      const init = new MethodDeclaration('', this);
      init.identifiers.push({
        id: 'init',
      });
      init.returnType = 'id';
      this.properties.push(alloc, init);
    }
    this.properties.forEach(property => {
      code.appendLine(property.generate());
    });
    code.endIndent().appendLine('}')
    code.endIndent().appendLine('}');
    if (this.type === 'interface') {
      code.appendLine().appendLine(`declare const ${this.identifier}: ${normalizeType(this.identifier)};`)
    }
    code.appendLine();
    return code.toString();
  }
}
