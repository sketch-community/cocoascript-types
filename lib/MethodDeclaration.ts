import { CodeGenerator } from './CodeGenerator';
import { normalizeType } from './helpers';
import { InterfaceDeclaration } from './InterfaceDeclaration';
import { Token } from './types';

interface Identifier {
  id: string;
  paramType?: string;
  paramName?: string;
}

const keywordBlacklist = ['unsigned', 'const', '_Nullable', '_Null_unspecified', '_Nonnull'];

export class MethodDeclaration {
  id: string;
  interfaceDecl: InterfaceDeclaration;
  identifiers: Identifier[] = [];
  returnType: string | null = null;

  constructor(id: string, interfaceDecl: InterfaceDeclaration) {
    this.id = id;
    this.interfaceDecl = interfaceDecl;
  }

  static initFromTokens(id: string, interfaceDecl: InterfaceDeclaration, tokens: Token[]) {
    let index = 0;
    const method = new MethodDeclaration(id, interfaceDecl);
    while (index < tokens.length) {
      if (
        (tokens[index].kind === 'typeIdentifier' && tokens[index].preciseIdentifier) ||
        tokens[index].kind === 'keyword'
      ) {
        method.returnType = tokens[index].text;
      }

      if (tokens[index].kind === 'internalParam' && tokens[index].text === '...') {
        const lastIndex = method.identifiers.length - 1;
        const lastId = method.identifiers[lastIndex];
        method.identifiers[lastIndex] = {
          ...lastId,
          paramName: `...${lastId.paramName}`,
          paramType: `Array<${normalizeType(lastId.paramType!)}>`,
        };
      }

      if (tokens[index].kind === 'identifier') {
        const id = tokens[index].text.replace(/:$/, '');
        let paramType;
        let paramName;
        if (tokens[index]) {
          while ((!paramName || !paramType) && tokens[index]) {
            if (
              tokens[index].kind === 'typeIdentifier' ||
              (tokens[index].kind === 'keyword' && !keywordBlacklist.includes(tokens[index].text))
            ) {
              paramType = tokens[index].text;
            }
            if (tokens[index].kind === 'internalParam') {
              paramName = tokens[index].text;
            }
            index += 1;
          }
        }
        method.identifiers.push({
          id,
          paramType,
          paramName,
        });
      }

      index += 1;
    }

    return method;
  }

  generate() {
    const code = new CodeGenerator();
    code.appendLine(`// ${this.id}`);
    code.appendLine(this.identifiers.map(i => i.id).join('_'));
    code.append('(');
    const params: string[] = [];
    const usedName: Set<string> = new Set();
    this.identifiers.forEach((identifier, i) => {
      if (identifier.paramName) {
        const name = usedName.has(identifier.paramName) ? `${identifier.paramName}${i}` : identifier.paramName;
        usedName.add(identifier.paramName);
        params.push(
          `${name}: ${normalizeType(
            identifier.paramType! === 'id' ? this.interfaceDecl.identifier : identifier.paramType!
          )}`
        );
      }
    });
    code.append(params.join(', '));
    code.append('):');
    let returnType = 'void';
    if (this.returnType === 'instancetype') {
      returnType = this.interfaceDecl.identifier;
    } else if (this.returnType === 'id') {
      returnType = this.interfaceDecl.identifier;
    } else if (this.returnType) {
      returnType = this.returnType;
    }
    code.append(normalizeType(returnType, { extend: false }));
    code.append(';');
    return code.toString();
  }
}
