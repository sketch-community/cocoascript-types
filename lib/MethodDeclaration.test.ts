import { getTokens } from './helpers';
import { InterfaceDeclaration } from './InterfaceDeclaration';
import { MethodDeclaration } from './MethodDeclaration';
import { APPLE_IDENTIFIER_PREFIX } from './constant';

describe('MethodDeclaration', () => {
  const nsview = require('../documentation/appkit/nsview.json');
  const interfaceDecl = InterfaceDeclaration.initFromTokens(
    `"${APPLE_IDENTIFIER_PREFIX}/documentation/appkit/nsview"`,
    {},
    getTokens(nsview)
  );

  function getDecl(docFile: string) {
    const doc = require(docFile);
    return MethodDeclaration.initFromTokens('fakeid', interfaceDecl, getTokens(doc));
  }

  it('simple method', () => {
    const decl = getDecl('../documentation/appkit/nsview/1483458-initwithframe.json');
    expect(decl.generate()).toMatchSnapshot();
  });

  it('multiple params', () => {
    const decl = getDecl('../documentation/appkit/nsview/1483640-addsubview.json');
    expect(decl.generate()).toMatchSnapshot();
  });

  it('with nullable', () => {
    const decl = getDecl('../documentation/appkit/nsworkspace/1526144-typeoffile.json');
    expect(decl.generate()).toMatchSnapshot();
  });

  it('duplicate params', () => {
    const decl = getDecl('../documentation/uikit/uistepper/1624072-dividerimageforleftsegmentstate.json');
    expect(decl.generate()).toMatchSnapshot();
  });

  it('unsigned params', () => {
    const decl = getDecl('../documentation/appkit/nsbitmapimagerep/1395538-initwithbitmapdataplanes.json');
    expect(decl.generate()).toMatchSnapshot();
  });
});
