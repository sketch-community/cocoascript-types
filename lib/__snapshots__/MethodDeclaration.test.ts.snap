// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MethodDeclaration duplicate params 1`] = `
"// fakeid
dividerImageForLeftSegmentState_rightSegmentState(state: UIControlState, state1: UIControlState):UIImage;"
`;

exports[`MethodDeclaration multiple params 1`] = `
"// fakeid
addSubview_positioned_relativeTo(view: NSView, place: NSWindowOrderingMode, otherView: NSView):void;"
`;

exports[`MethodDeclaration simple method 1`] = `
"// fakeid
initWithFrame(frameRect: NSRect):NSView;"
`;

exports[`MethodDeclaration unsigned params 1`] = `
"// fakeid
initWithBitmapDataPlanes_pixelsWide_pixelsHigh_bitsPerSample_samplesPerPixel_hasAlpha_isPlanar_colorSpaceName_bitmapFormat_bytesPerRow_bitsPerPixel(planes: string, width: NSInteger, height: NSInteger, bps: NSInteger, spp: NSInteger, alpha: BOOL, isPlanar: BOOL, colorSpaceName: NSColorSpaceName, bitmapFormat: NSBitmapFormat, rBytes: NSInteger, pBits: NSInteger):NSView;"
`;

exports[`MethodDeclaration with nullable 1`] = `
"// fakeid
typeOfFile_error(absoluteFilePath: string | cocoascript.NSString, outError: NSError):NSString;"
`;
